src/gluonts/torch/distributions/mqf2.py

```diff
@@ -0,0 +1,395 @@
+    Distribution,
+    DistributionOutput,
+)
+
+from torch.distributions import (
+    AffineTransform,
+    TransformedDistribution,
+)
+
+from torch.distributions.normal import Normal
+
+
+class MQF2Distribution(Distribution):
+    r"""
+    Distribution class for the model MQF2 proposed in the paper
+    ``Multivariate Quantile Function Forecaster``
+    by Kan, Aubet, Januschowski, Park, Benidis, Ruthotto, Gasthaus
+
+    Parameters
+    ----------
+    picnn
+        A SequentialNet instance of a partially input convex neural network (picnn)
+    hidden_state
+        hidden_state obtained by unrolling the RNN encoder
+        shape = (batch_size, context_length, hidden_size) in training
+        shape = (batch_size, hidden_size) in inference
+    prediction_length
+        Length of the prediction horizon
+    is_energy_score
+        If True, use energy score as objective function
+        otherwise use maximum likelihood as objective function (normalizing flows)
+    es_num_samples
+        Number of samples drawn to approximate the energy score
+    beta
+        Hyperparameter of the energy score (power of the two terms)
+    threshold_input
+        Clamping threshold of the (scaled) input when maximum likelihood is used as objective function
+        this is used to make the forecaster more robust to outliers in training samples
+    validate_args
+        Sets whether validation is enabled or disabled
+        For more details, refer to the descriptions in torch.distributions.distribution.Distribution
+    """
+
+    def __init__(
+        self,
+        picnn: torch.nn.Module,
+        hidden_state: torch.Tensor,
+        prediction_length: int,
+        is_energy_score: bool = True,
+        es_num_samples: int = 50,
+        beta: float = 1.0,
+        threshold_input: float = 100.0,
+        validate_args: bool = False,
+    ) -> None:
+
+        self.picnn = picnn
+        self.hidden_state = hidden_state
+        self.prediction_length = prediction_length
+        self.is_energy_score = is_energy_score
+        self.es_num_samples = es_num_samples
+        self.beta = beta
+        self.threshold_input = threshold_input
+
+        super().__init__(
+            batch_shape=self.batch_shape, validate_args=validate_args
+        )
+
+        self.context_length = (
+            self.hidden_state.shape[-2]
+            if len(self.hidden_state.shape) > 2
+            else 1
+        )
+        self.numel_batch = MQF2Distribution.get_numel(self.batch_shape)
+
+        zero = torch.tensor(
+            0, dtype=hidden_state.dtype, device=hidden_state.device
+        )
+        one = torch.ones_like(zero)
+        self.standard_normal = Normal(zero, one)
+
+    def loss(self, z: torch.Tensor) -> torch.Tensor:
+        if self.is_energy_score:
+            return self.energy_score(z)
+        else:
+            return -self.log_prob(z)
+
+    def unfold_observations(self, z: torch.Tensor) -> torch.Tensor:
```

Comment:
In the end of the day, it is more of a `reshape` than `unfold`. Maybe adapt the name?

Reply:
It's actually both. It is actually sliding through the tensor and stack the sliding view together. I will rename it as stack_sliding_view so that it is more descriptive.